Fix Issue 20951 - Don't use excess precision when parsing single- and double-precision literals

kinke opened the pull request.

kinke committed:
Fix Issue 20951 - Don't use excess precision when parsing single- and double-precision literals

kinke commented:
Needs some Phobos adaptations in `std.math` and `std.conv`, plus looking into the failure for the newly added test on Windows hosts (imprecise `strtold_dm()`?).

Also breaks a few buildkite projects.

9il commented:
The excess precision or downcase precision still should be used in some cases.

Example that should work:
```
real a == -0x8.80000000000019fp-514L;
real b == -0x8.80000000000019fp-514;
assert(a == b);
assert(b == -0x8.80000000000019fp-514);
```

The literal when no `f` trailing `L` is defined should take the type of information from the other side of the `==` or `=` if any.

This is how it works in C.

kinke commented:
> This is how it works in C.

Not according to godbolt (for C++): https://cpp.godbolt.org/z/-TRDU8

```
#include <cassert>

int main()
{
    long double a = -0x8.80000000000019fp-514L;
    long double b = -0x8.80000000000019fp-514;
    assert(a == b); // fails
    assert(a == -0x8.80000000000019fp-514); // fails
}
```

9il commented:
> Not according to godbolt (for C++, with both gcc and clang): https://cpp.godbolt.org/z/-TRDU8

My bad. Please ignore my previous comment then.

WalterBright commented on src/dmd/hdrgen.d:
```diff
@@ -1888,41 +1888,36 @@ public:
-        const(size_t) BUFFER_LEN = value.sizeof * 3 + 8 + 1 + 1;
+        enum BUFFER_LEN = value.sizeof * 3 + 8 + 1 + 1;
         char[BUFFER_LEN] buffer;
+
```
buffer should be initialized to void.

WalterBright commented on src/dmd/root/ctfloat.d:
```diff
@@ -167,26 +167,43 @@ extern (C++) struct CTFloat
 
     @system
-    static real_t parse(const(char)* literal, bool* isOutOfRange = null)
+    extern (D) private static real_t parse(alias parseFn)(const(char)* literal, bool* isOutOfRange)
```
`isOutOfRange` should be `out bool`.

kinke replied:
That only complicates matters for the actual functions forwarding to this - they all take an optional pointer.

WalterBright replied:
Are there any that do this? Generally, I've been trying to move away from this old C style.

kinke replied:
Well, it's the usual solution for an optional `out` param. Parsing these literals occurs in 2 places here upstream - the lexer, which cares about any overflow, and hdrgen, which doesn't care about it because it's just used to determine whether the `%g` decimal string can be losslessly re-parsed to exactly the same value. In LDC, we have a bunch of additional calls, and IIRC, we don't care about the overflow there either.

9il commented:
`mir-algorithm` has been updated. The bug was the missing trailing  `L` for real literal in `==` expression.

WalterBright commented on src/dmd/root/ctfloat.d:
```diff
@@ -167,26 +167,43 @@ extern (C++) struct CTFloat
         {
             auto save = __locale_decpoint;
             __locale_decpoint = ".";
+            scope(exit) __locale_decpoint = save;
```
Leave the old way, as this way causes an exception handling frame to be built, which is slow.

kinke replied:
`parseFn` should be nothrow in all cases, hence no `finally` should be necessary.

WalterBright replied:
The operative word is "should be". Check and see.

kinke replied:
Right; added `nothrow` to the `parse()` function.

9il commented:
relevant issue https://issues.dlang.org/show_bug.cgi?id=20972

PetarKirov commented:
@kinke @ibuclaw @WalterBright what's the status of this PR? What needs to be done before it's merged?

kinke commented:
Well, as Ilya pointed out in the forum IIRC, this won't work at all with the DigitalMars C runtime (because it's using `strtold` for `strtod` and `strtof`), i.e., any DMD on Windows compiled with a DMD host compiler and `-m32` (but `-m64` and `-m32mscoff` probably work).

And as the Buildkite failures show, this will probably have to be opt-in via `-preview`.

WalterBright commented on src/dmd/root/ctfloat.d:
```diff
@@ -167,26 +167,43 @@ extern (C++) struct CTFloat
+    static real_t parseReal(const(char)* literal, bool* isOutOfRange = null)
+    {
+        version(CRuntime_Microsoft) // Microsoft's strtold is for 64-bit `long double`
+            alias parseFn = strtold_dm;
```
Seems to me if strtold_dm is always used, it will avoid the off-by-last-bit in dmc's runtime implementation of strtold.

WalterBright commented:
Seems to me that the most reliable way to nail this is to:

1. use strtold_dm for all 80 bit real conversions

2. take a cue from strtold_dm and implement strtof_dm and strtod_dm that will work constantly and correctly on all platforms

WalterBright commented:
It appears that strtold_dm has the same problem as strtold identified here https://issues.dlang.org/show_bug.cgi?id=20972

Therefore, I propose that strtold_dm() be fixed first.

kinke commented:
> take a cue from strtold_dm and implement strtof_dm and strtod_dm that will work consistently and correctly on all platforms

* LDC and GDC don't need any of this.
* All C runtimes except DigitalMars should work fine.

WalterBright reviewed (approved):
@kinke in that case, I agree with you. I approve of this PR, although for the time being the correctness tests should be disabled for the dmd on Win32/Win64 builds.

kinke committed:
[squash] Slightly revise

kinke committed:
Use strtold_dm() instead of DMC strtold() for parsing real literals

I.e., only when using a DMD host compiler *and* -m32.

This is a preparation for a fix for issue #20972 - fixing strtold_dm()
will fix parsed real literals on Windows in general (and not only when
linking DMD against the MSVC runtime).

wilzbach commented:
> I clearly propose to drop DMC support, at least for the compiler itself, but having to maintain it in druntime and Phobos isn't worth it either IMO.

Yeah, AFAICT everyone active here (except Walter) is on your side.

kinke commented:
> in that case, I agree with you. although for the time being the correctness tests should be disabled for the dmd on Win32/Win64 builds.

Thx & done. The remaining blockers are Phobos and a few Buildkite projects, if we don't want to make this opt-in/out.

kinke commented:
I've taken care of Phobos (well, at least for Linux x64), but I'm definitely not going to tackle the Buildkite projects.

nu8 commented:
https://github.com/ldc-developers/ldc/issues/3561

thewilsonator commented:
cc @BlackEdder (ggplotd) @CyberShadow (ae) @ibuclaw (ocean) you all have projects that fail unittests as a result of this PR

CyberShadow commented:
Here is a reduction of the ae failure:

```d
import std.conv;

void main()
{
	real a = 123e-2;
	real b = "123e-2".to!real;
	assert(a == b);
}
```

This works with DMD master but not with this pull request.

From reading the discussion above I infer that this breakage is intentional. However, this seems like a sneaky and subtle change, and simply going forward with this feels like a bad idea.

The failures that we see on BuildKite are likely only a small percentage of code that will be affected by this. Even putting aside the usual fact that not all D code is on BuildKite -- more importantly, this is much more likely to manifest as silent changes in results which are not covered by automatic tests.

We generally try to not silently change the behavior of code, and this should be no exception. Would it be possible to emit a warning for such cases? I.e., add a flag to CTFE `double` values which indicates loss of precision compared to if the literal had been parsed as a `real` instead, and then, if a value with such a flag on is cast to a `real`, emit a warning. Users can then choose to add `L` to preserve the old behavior or explicitly cast to `double` (which would remove the flag) to silence the warning.

The warning would be useful regardless of the breaking-change aspect of this. The line `real a = 123e-2;` does not do what readers might think it does, so it would be a good addition either way.

a8nJ commented:
https://github.com/dlang/dmd/pull/11387
https://github.com/dlang/phobos/pull/7591
https://github.com/ldc-developers/ldc/pull/3498

kinke commented:
> We generally try to not silently change the behavior of code, and this should be no exception. Would it be possible to emit a warning for such cases? I.e., add a flag to CTFE double values which indicates loss of precision compared to if the literal had been parsed as a real instead, and then, if a value with such a flag on is cast to a real, emit a warning.

FWIW, I've looked into this back then, and unfortunately, the information about the literal precision is lost after the parsing stage, as IIRC, all FP literals (`TOK.float{32,64,80}`) are normalized to `TOK.float64`, presumably to simplify/accelerate the numerous is-FP-literal checks.

WalterBright commented:
Even 80 bit literals are normalized to 64? When did that happen?

kinke commented:
The value is left alone, but `tok` is `TOK.float64`: https://github.com/dlang/dmd/blob/96afc26b2c9ff4694b9a1bef7399b95b3582261d/src/dmd/expression.d#L2091
